src/platform/i386/hpet.c

```diff
@@ -142,86 +143,86 @@ timer_calibration(void)
 		tot += cycle - prev;
 		apic_tot += (apic_prev - apic_curr);
 	}
-	if (cnt >= TIMER_CALIBRATION_ITER) {
-		assert(hpetcyc_per_tick);
-		timer_calibration_init = 0;
-		cycles_per_tick        = (unsigned long)(tot / TIMER_CALIBRATION_ITER);
-		assert(cycles_per_tick > hpetcyc_per_tick);
+	if (cnt >= HPET_CALIBRATION_ITER) {
+		assert(hpet_hpetcyc_per_tick);
+		hpet_calibration_init = 0;
+		hpet_cpucyc_per_tick  = (unsigned long)(tot / HPET_CALIBRATION_ITER);
+		assert(hpet_cpucyc_per_tick > hpet_hpetcyc_per_tick);
 
 		if (lapic_timer_calib_init) {
 			u32_t cycs_to_apic_ratio = 0, apic_cycs_per_tick = 0;
 
-			apic_cycs_per_tick = apic_tot / TIMER_CALIBRATION_ITER;
+			apic_cycs_per_tick = apic_tot / HPET_CALIBRATION_ITER;
 			assert(apic_cycs_per_tick);
 
-			cycs_to_apic_ratio = cycles_per_tick / apic_cycs_per_tick;
+			cycs_to_apic_ratio = hpet_cpucyc_per_tick / apic_cycs_per_tick;
 			lapic_timer_calibration(cycs_to_apic_ratio);
 		}
 
 		/* Possibly significant rounding error here.  Bound by the factor */
-		timer_cycles_per_hpetcyc = (TIMER_ERROR_BOUND_FACTOR * cycles_per_tick) / hpetcyc_per_tick;
+		hpet_cpucyc_per_hpetcyc = (HPET_ERROR_BOUND_FACTOR * hpet_cpucyc_per_tick) / hpet_hpetcyc_per_tick;
 		printk("Timer calibrated:\n\tCPU cycles per HPET tick: %ld\n\tHPET ticks in %d us: %ld\n",
-		       timer_cycles_per_hpetcyc / TIMER_ERROR_BOUND_FACTOR, TIMER_DEFAULT_US_INTERARRIVAL,
-		       hpetcyc_per_tick);
+		       hpet_cpucyc_per_hpetcyc / HPET_ERROR_BOUND_FACTOR, HPET_DEFAULT_PERIOD_US,
+		       hpet_hpetcyc_per_tick);
 
-		timer_disable(TIMER_PERIODIC);
-		timer_disable(TIMER_PERIODIC);
+		hpet_disable(HPET_PERIODIC);
+		hpet_disable(HPET_PERIODIC);
 	}
 	cnt++;
 }
 
 int
 chal_cyc_usec(void)
 {
-	return cycles_per_tick / TIMER_DEFAULT_US_INTERARRIVAL;
+	return hpet_cpucyc_per_tick / HPET_DEFAULT_PERIOD_US;
 }
 
 int
-periodic_handler(struct pt_regs *regs)
+hpet_periodic_handler(struct pt_regs *regs)
 {
 	int preempt = 1;
 
-	if (unlikely(timer_calibration_init)) timer_calibration();
+	if (unlikely(hpet_calibration_init)) hpet_calibration();
 
-	ack_irq(HW_PERIODIC);
-	preempt = cap_hw_asnd(&hw_asnd_caps[HW_PERIODIC], regs);
-	HPET_INT_ENABLE(TIMER_PERIODIC);
+	lapic_ack();
+	preempt = cap_hw_asnd(&hw_asnd_caps[HW_HPET_PERIODIC], regs);
+	HPET_INT_ENABLE(HPET_PERIODIC);
 
 	return preempt;
 }
 
-extern int timer_process(struct pt_regs *regs);
-
 int
-oneshot_handler(struct pt_regs *regs)
+hpet_oneshot_handler(struct pt_regs *regs)
 {
 	int preempt = 1;
 
-	ack_irq(HW_ONESHOT);
-	preempt = timer_process(regs);
-	HPET_INT_ENABLE(TIMER_ONESHOT);
+	assert(!hpet_calibration_init);
+
+	lapic_ack();
+	preempt = cap_hw_asnd(&hw_asnd_caps[HW_HPET_ONESHOT], regs);
+	HPET_INT_ENABLE(HPET_ONESHOT);
```

Comment:
I'm not sure either, I'll read the HPET specification again.
But I'm just doing exactly what was done before and it was working before. ;)

Reply:
I looked at the specification for this.
For some reason we program the HPET to be "level-triggered" and the specification mentions that the software is required to clear the interrupt by setting a flag. So we do that every time.

It was good to revisit the specification because I found something useful that we can do with enabling IOAPIC, it seems we enable LEGACY INTERRUPT ROUTING support in the HPET right now which we can get rid of and configure the HPET to use specific lines in APIC mode. I'll add that in as an issue and look at it later.